File: upstream/workers/hs-ledger/nenrin/resume-v1/resume_v1.mjs
// resume_v1.mjs : JS port of resume_v1.py. Same rules, same rejection codes,
// byte-identical canonical form (sorted keys, no whitespace, raw UTF-8).
//
// No node imports on purpose: the ledger worker bundles this file and passes its
// own Web Crypto hasher. sha256Hex is injected (may be sync or async), so the
// whole assembly is async. node users pass createHash (see resume_cli.mjs).
// M4: two implementations, one sha.

export const ALLOWED_OUTCOME = new Set(["PASS", "FAIL"]);
export const FORBIDDEN_SCORE_KEYS = new Set(["score", "rating", "stars", "points", "rank", "grade", "trust_score"]);

export function canonical(v) {
  if (v === null || v === undefined) return "null";
  if (Array.isArray(v)) return "[" + v.map(canonical).join(",") + "]";
  if (typeof v === "object") {
    return "{" + Object.keys(v).sort().map((k) => JSON.stringify(k) + ":" + canonical(v[k])).join(",") + "}";
  }
  return JSON.stringify(v);
}

export class Reject extends Error {
  constructor(code, why) { super(code + ": " + why); this.code = code; this.why = why; }
}

// python dict.get(k): missing -> null, present falsy values kept as they are
const g = (o, k) => (o && o[k] !== undefined ? o[k] : null);
// python truthiness (None, False, 0, "", [], {} are falsy)
const pyTruthy = (v) => !(v === null || v === undefined || v === false || v === 0 || v === "" ||
  (Array.isArray(v) && v.length === 0) || (typeof v === "object" && !Array.isArray(v) && Object.keys(v).length === 0));
const pyOr = (a, b) => (pyTruthy(a) ? a : b);

const ISO_RE = /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})Z$/;
const LEDGER_RE = /^(\d{4})-(\d{2})-(\d{2}) (\d{2}):(\d{2})(?::(\d{2}))? UTC$/;
// UTC seconds for 'YYYY-MM-DDTHH:MM:SSZ' (walk records) or 'YYYY-MM-DD HH:MM[:SS] UTC'
// (the ledger's block_time, written by the stamping runner). null if neither.
export function toEpoch(t) {
  if (typeof t !== "string") return null;
  const m = ISO_RE.exec(t) || LEDGER_RE.exec(t);
  if (!m) return null;
  const y = +m[1], mo = +m[2], d = +m[3], h = +m[4], mi = +m[5], s = +(m[6] || 0);
  if (mo < 1 || mo > 12 || d < 1 || d > 31 || h > 23 || mi > 59 || s > 59) return null;
  const ms = Date.UTC(y, mo - 1, d, h, mi, s);
  // reject calendar overflow (e.g. Feb 30) the way python's datetime does
  const chk = new Date(ms);
  if (chk.getUTCMonth() !== mo - 1 || chk.getUTCDate() !== d) return null;
  return Math.floor(ms / 1000);
}
function within(t, now, days) {
  // python: (to_epoch(now) - to_epoch(t)) // 86400 <= days ; floor division
  return Math.floor((toEpoch(now) - toEpoch(t)) / 86400) <= days;
}

function scanForbidden(obj, path = "$") {
  if (Array.isArray(obj)) { obj.forEach((v, i) => scanForbidden(v, path + "[" + i + "]")); return; }
  if (obj && typeof obj === "object") {
    for (const k of Object.keys(obj)) {
      if (FORBIDDEN_SCORE_KEYS.has(String(k).toLowerCase())) throw new Reject("score_injection", "forbidden score key '" + k + "' at " + path);
      scanForbidden(obj[k], path + "." + k);
    }
  }
}

function verdictOf(rec) {
  const v = rec.verdict;
  if (v && typeof v === "object" && !Array.isArray(v)) return [g(v, "outcome"), g(v, "ok"), g(v, "n_pass"), g(v, "n_total")];
  return [g(rec, "outcome"), null, null, null];
}

export async function checkMeasurement(m, sha256Hex) {
  const rc = m.record_canonical, claimed = m.record_sha256;
  if (typeof rc !== "string" || typeof claimed !== "string") throw new Reject("self_asserted", "measurement carries no record bytes to authenticate");
  if ((await sha256Hex(rc)) !== claimed) throw new Reject("orphan_record", "record_sha256 does not recompute from record_canonical bytes");
  let rec;
  try { rec = JSON.parse(rc); } catch (_e) { throw new Reject("self_asserted", "record_canonical is not JSON"); }
  if (!rec || typeof rec !== "object" || Array.isArray(rec) || rec.schema !== "jidec-path-v1") throw new Reject("self_asserted", "record is not a jidec-path-v1 measurement");
  scanForbidden(rec);
  const [outcome, ok, nPass, nTotal] = verdictOf(rec);
  if (!ALLOWED_OUTCOME.has(outcome)) throw new Reject("score_injection", "verdict.outcome must be a category, got " + JSON.stringify(outcome));
  if (ok !== null && Boolean(ok) !== (outcome === "PASS")) throw new Reject("verdict_inconsistent", "verdict.ok disagrees with verdict.outcome");
  const w = rec.witness;
  if (!w || typeof w !== "object" || Array.isArray(w) || !pyTruthy(w.name) || !pyTruthy(w.vantage)) throw new Reject("self_asserted", "measurement has no witness{name,vantage}");
  const anchor = m.anchor || {};
  const blockTime = g(anchor, "block_time");
  const measuredAt = pyOr(pyOr(g(rec, "walked_at"), g(rec, "measured_at")), g(rec, "first_instant"));
  if (!pyTruthy(blockTime) || !pyTruthy(measuredAt)) throw new Reject("coordinate_chosen_by_prover", "no anchor block_time to bound the measurement time");
  const tM = toEpoch(measuredAt), tB = toEpoch(blockTime);
  if (tM === null || tB === null) throw new Reject("coordinate_chosen_by_prover", "measurement or anchor time is not a recognised UTC timestamp");
  if (tM > tB) throw new Reject("coordinate_chosen_by_prover", "walked_at is after the anchoring block (postdated)");
  return [rec, outcome, w, measuredAt, nPass, nTotal];
}

async function copyRing(r, discrepancies, sha256Hex) {
  const rc = r.record_canonical, claimed = r.record_sha256;
  if (typeof rc !== "string" || (await sha256Hex(rc)) !== claimed) throw new Reject("orphan_record", "ring record_sha256 does not recompute (M1)");
  const rr = JSON.parse(rc);
  scanForbidden(rr);
  for (const d of pyOr(g(rr, "discrepancies"), [])) discrepancies.push({ record_sha256: claimed, disc: d });
  let counts = g(rr, "counts");
  if (!pyTruthy(counts)) { counts = {}; for (const k of ["instants_reached", "instants_sampled"]) if (k in rr) counts[k] = rr[k]; }
  return {
    month: pyOr(g(rr, "month"), g(rr, "from")),
    endpoint: g(rr, "endpoint"),
    counts,
    determinism: g(rr, "determinism"),
    derived: g(rr, "derived"),
    digest: g(rr, "digest"),
    ledger_n: g(r, "source_ledger_n"),
  };
}

export async function assembleResume(permaId, endpoint, agentCardUrl, measurements, opts = {}) {
  const sha256Hex = opts.sha256Hex;
  if (typeof sha256Hex !== "function") throw new Error("assembleResume: opts.sha256Hex (string -> hex, sync or async) is required");
  const rings = opts.rings || [], agreements = opts.agreements || [];
  const periodDays = opts.period_days === undefined ? 30 : opts.period_days;
  const now = opts.now === undefined ? null : opts.now;
  const outMeas = [], discrepancies = [], counts = { PASS: 0, FAIL: 0 };
  const names = new Set(), vantages = new Set(), times = [];
  for (const m of measurements) {
    const [rec, outcome, w, measuredAt, nPass, nTotal] = await checkMeasurement(m, sha256Hex);
    counts[outcome] += 1; names.add(w.name); vantages.add(w.vantage); times.push(measuredAt);
    for (const d of pyOr(g(rec, "discrepancies"), [])) discrepancies.push({ record_sha256: m.record_sha256, disc: d });
    const anchor = m.anchor || {};
    outMeas.push({
      measured_at: measuredAt,
      record_sha256: m.record_sha256,
      outcome,
      n_pass: nPass,
      n_total: nTotal,
      base: g(rec, "base"),
      purpose: g(rec, "purpose"),
      witness: { name: w.name, vantage: w.vantage, key_url: g(w, "key_url") },
      record_url: g(m, "record_url"),
      anchor: { bitcoin_block: g(anchor, "bitcoin_block"), block_time: g(anchor, "block_time"), ots: g(anchor, "ots"), batch_sha256: g(anchor, "batch_sha256") },
      source_ledger_n: g(m, "source_ledger_n"),
    });
  }
  const outRings = [];
  for (const r of rings) outRings.push(await copyRing(r, discrepancies, sha256Hex));
  const last = times.length ? times.reduce((a, b) => (a > b ? a : b)) : null;
  const oldest = times.length ? times.reduce((a, b) => (a < b ? a : b)) : null;
  const currentNow = Boolean(last && now && toEpoch(last) !== null && toEpoch(now) !== null && within(last, now, periodDays));
  const resume = {
    schema: "nenrin-resume-v1",
    perma_id: permaId, measured_endpoint: endpoint, agent_card_url: agentCardUrl,
    counts,
    witness_diversity: { distinct_names: names.size, distinct_vantages: vantages.size },
    measurements: outMeas, discrepancies, rings: outRings,
    agreements: agreements.map((a) => ({ record_sha256: g(a, "record_sha256"), ledger_n: g(a, "source_ledger_n") })),
    freshness: { last_measured: last, oldest_measurement: oldest, current_now: currentNow, period_days: periodDays },
  };
  resume.resume_sha256 = await sha256Hex(canonical(resume));
  return resume;
}

File: upstream/workers/hs-ledger/nenrin/resume-v1/resume_v1.py
"""resume_v1.py : NENRIN Resume v1 reference assembler (offline, deterministic).

The resume makes no new claim. It assembles existing jidec-path-v1 witness-walk
records (a2a_conduct_walk.py shape: verdict{ok,outcome,n_pass,n_total}, walked_at,
base, witness{name,vantage}, nodes[]) under one identity and refuses, by
construction, any line that is not authenticated by its own bytes. Same
canonicalization as make_ring.py so the bytes match the worker.

Three laws enforced here:
  1. no self-asserted line  (every line comes from a record whose bytes hash to its sha)
  2. discrepancies are copied verbatim, never dropped
  3. no score is ever emitted (counts and links only; PASS/FAIL and n_pass/n_total are counts)
"""
import json, hashlib, re
from datetime import datetime, timezone

ALLOWED_OUTCOME = {"PASS", "FAIL"}  # witness-walk vocabulary (verdict.outcome)
FORBIDDEN_SCORE_KEYS = {"score", "rating", "stars", "points", "rank", "grade", "trust_score"}


def canonical(obj):
    return json.dumps(obj, ensure_ascii=False, sort_keys=True, separators=(",", ":"))


def sha256_hex(s):
    return hashlib.sha256(s.encode("utf-8")).hexdigest()


class Reject(Exception):
    def __init__(self, code, why):
        self.code = code
        self.why = why
        super().__init__(code + ": " + why)


_ISO = re.compile(r"^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})Z$")
_LEDGER = re.compile(r"^(\d{4})-(\d{2})-(\d{2}) (\d{2}):(\d{2})(?::(\d{2}))? UTC$")


def to_epoch(t):
    """UTC seconds for 'YYYY-MM-DDTHH:MM:SSZ' (walk records) or 'YYYY-MM-DD HH:MM[:SS] UTC'
    (the ledger's block_time, written by the stamping runner). None if neither."""
    if not isinstance(t, str):
        return None
    m = _ISO.match(t) or _LEDGER.match(t)
    if not m:
        return None
    y, mo, d, h, mi, sec = m.groups()
    try:
        return int(datetime(int(y), int(mo), int(d), int(h), int(mi), int(sec or 0), tzinfo=timezone.utc).timestamp())
    except ValueError:
        return None


def _within(t, now, days):
    # floor of whole days, same as JS Math.floor((now - t) / 86400000)
    return (to_epoch(now) - to_epoch(t)) // 86400 <= days


def _scan_forbidden(obj, path="$"):
    if isinstance(obj, dict):
        for k, v in obj.items():
            if k.lower() in FORBIDDEN_SCORE_KEYS:
                raise Reject("score_injection", "forbidden score key '" + k + "' at " + path)
            _scan_forbidden(v, path + "." + str(k))
    elif isinstance(obj, list):
        for i, v in enumerate(obj):
            _scan_forbidden(v, path + "[" + str(i) + "]")


def _verdict(rec):
    """verdict{ok,outcome,n_pass,n_total} (walk shape). Legacy: top-level outcome, no ok."""
    v = rec.get("verdict")
    if isinstance(v, dict):
        return v.get("outcome"), v.get("ok"), v.get("n_pass"), v.get("n_total")
    return rec.get("outcome"), None, None, None


def check_measurement(m):
    """Authenticate one measurement. Returns (rec, outcome, witness, measured_at, n_pass, n_total) or raises Reject."""
    rc = m.get("record_canonical")
    claimed = m.get("record_sha256")
    if not isinstance(rc, str) or not isinstance(claimed, str):
        raise Reject("self_asserted", "measurement carries no record bytes to authenticate")
    if sha256_hex(rc) != claimed:
        raise Reject("orphan_record", "record_sha256 does not recompute from record_canonical bytes")
    try:
        rec = json.loads(rc)
    except Exception:
        raise Reject("self_asserted", "record_canonical is not JSON")
    if not isinstance(rec, dict) or rec.get("schema") != "jidec-path-v1":
        raise Reject("self_asserted", "record is not a jidec-path-v1 measurement")
    _scan_forbidden(rec)
    outcome, ok, n_pass, n_total = _verdict(rec)
    if outcome not in ALLOWED_OUTCOME:
        raise Reject("score_injection", "verdict.outcome must be a category in " + str(sorted(ALLOWED_OUTCOME)) + ", got " + repr(outcome))
    if ok is not None and bool(ok) != (outcome == "PASS"):
        raise Reject("verdict_inconsistent", "verdict.ok disagrees with verdict.outcome")
    w = rec.get("witness")
    if not isinstance(w, dict) or not w.get("name") or not w.get("vantage"):
        raise Reject("self_asserted", "measurement has no witness{name,vantage}")
    anchor = m.get("anchor") or {}
    block_time = anchor.get("block_time")
    measured_at = rec.get("walked_at") or rec.get("measured_at") or rec.get("first_instant")
    if not block_time or not measured_at:
        raise Reject("coordinate_chosen_by_prover", "no anchor block_time to bound the measurement time")
    t_m, t_b = to_epoch(measured_at), to_epoch(block_time)
    if t_m is None or t_b is None:
        raise Reject("coordinate_chosen_by_prover", "measurement or anchor time is not a recognised UTC timestamp")
    if t_m > t_b:
        raise Reject("coordinate_chosen_by_prover", "walked_at is after the anchoring block (postdated)")
    return rec, outcome, w, measured_at, n_pass, n_total


def _copy_ring(r, discrepancies):
    rc = r.get("record_canonical")
    claimed = r.get("record_sha256")
    if not isinstance(rc, str) or sha256_hex(rc) != claimed:
        raise Reject("orphan_record", "ring record_sha256 does not recompute (M1)")
    rr = json.loads(rc)
    _scan_forbidden(rr)
    for d in (rr.get("discrepancies") or []):
        discrepancies.append({"record_sha256": claimed, "disc": d})
    return {
        "month": rr.get("month") or rr.get("from"),
        "endpoint": rr.get("endpoint"),
        "counts": rr.get("counts") or {k: rr.get(k) for k in ("instants_reached", "instants_sampled") if k in rr},
        "determinism": rr.get("determinism"),
        "derived": rr.get("derived"),
        "digest": rr.get("digest"),
        "ledger_n": r.get("source_ledger_n"),
    }


def assemble_resume(perma_id, endpoint, agent_card_url, measurements,
                    rings=None, agreements=None, period_days=30, now=None):
    rings = rings or []
    agreements = agreements or []
    out_meas = []
    discrepancies = []
    counts = {"PASS": 0, "FAIL": 0}
    names = set()
    vantages = set()
    times = []
    for m in measurements:
        rec, outcome, w, measured_at, n_pass, n_total = check_measurement(m)
        counts[outcome] += 1
        names.add(w["name"])
        vantages.add(w["vantage"])
        times.append(measured_at)
        for d in (rec.get("discrepancies") or []):
            discrepancies.append({"record_sha256": m["record_sha256"], "disc": d})
        anchor = m.get("anchor") or {}
        out_meas.append({
            "measured_at": measured_at,
            "record_sha256": m["record_sha256"],
            "outcome": outcome,
            "n_pass": n_pass,
            "n_total": n_total,
            "base": rec.get("base"),
            "purpose": rec.get("purpose"),
            "witness": {"name": w["name"], "vantage": w["vantage"], "key_url": w.get("key_url")},
            "record_url": m.get("record_url"),
            "anchor": {"bitcoin_block": anchor.get("bitcoin_block"), "block_time": anchor.get("block_time"), "ots": anchor.get("ots"), "batch_sha256": anchor.get("batch_sha256")},
            "source_ledger_n": m.get("source_ledger_n"),
        })
    out_rings = [_copy_ring(r, discrepancies) for r in rings]
    last = max(times) if times else None
    oldest = min(times) if times else None
    current_now = bool(last and now and to_epoch(last) is not None and to_epoch(now) is not None and _within(last, now, period_days))
    resume = {
        "schema": "nenrin-resume-v1",
        "perma_id": perma_id,
        "measured_endpoint": endpoint,
        "agent_card_url": agent_card_url,
        "counts": counts,
        "witness_diversity": {"distinct_names": len(names), "distinct_vantages": len(vantages)},
        "measurements": out_meas,
        "discrepancies": discrepancies,
        "rings": out_rings,
        "agreements": [{"record_sha256": a.get("record_sha256"), "ledger_n": a.get("source_ledger_n")} for a in agreements],
        "freshness": {"last_measured": last, "oldest_measurement": oldest, "current_now": current_now, "period_days": period_days},
    }
    resume["resume_sha256"] = sha256_hex(canonical(resume))
    return resume
